README.md
###Install instructions
Since package 'cvalues' uses over 30 other R packages its is suggested installing by using next command:
 

     devtools::install_github("dchristop/cvalues", dependencies=TRUE)

 
###Data available
The data created for World Values Survey wave six are available by using:

     data("FullCountryDataFrameForPlots")
     
     data("Final_W6_Archetypes")
     
     data("W6")
    

NEWS.md
## What's new in version 0.0.0.9002 released Feb 17, 2018

### Adding data *centroids* for central long-lat coordiantes of 241 countries

### Modifying function *mds\_country\_mixtures* in order to keep full rows of countries for plots

## What's new in version 0.0.0.9001 released Feb 13, 2018

### Fixing bug from package *rnaturalearth* data *countries110* issue

### Fixing note in compilation process for parallel variables in *foreach* package

### Adding *INSEAD Fontainebleau France* in author list with roles *cph* and *fnd*


R/data.R
#' A full data set created for the 60 countries of World Value Survey Wave 6
#'
#' @format A data frame with 60 rows and 10 variables:
#' \describe{
#' \item{name}{ country short name} 
#' \item{ISOAlpha2}{ country ISOAlpha2 code} 
#' \item{ISOAlpha3}{ country ISOAlpha3 code} 
#' \item{long}{ Longitude of country centroid} 
#' \item{lat}{ Latitude of country centroid} 
#' \item{D1}{ Dimension 1 of MDS} 
#' \item{D2}{ Dimension 2 of MDS} 
#' \item{D3}{ Dimension 3 of MDS} 
#' \item{Retreater}{ Country percentage for Retreaters} 
#' \item{Social_Hedonist}{ Country percentage for Social Hedonists} 
#' \item{Social_Conservative}{ Country percentage for Social Conservative} 
#' \item{Maximalist}{ Country percentage for Maximalists} 
#' \item{Social_Innovator}{ Country percentage for Social Innovator} 
#' \item{HH}{ HH Index} 
#' \item{NHH}{ NHH Index} 
#' \item{MI}{ MI Index} 
#' \item{GDP_Constant_$}{ GDP Constant $  Index} 
#' \item{Economic_Freedom}{ Economic Freedom  Index} 
#' \item{Political_Rights}{ Political Rights  Index} 
#' \item{Civil_Liberties}{ Civil Liberties  Index} 
#' \item{Overall_Freedom}{ Overall Freedom  Index} 
#' \item{Global_Competitiveness}{ Global Competitiveness  Index} 
#' \item{Human_Development}{ Human Development  Index} 
#' \item{Global_Innovation}{ Global Innovation  Index} 
#' \item{Rule_of_Law_1}{ Rule of Law 1  Index} 
#' \item{Control_of_corruption}{ Control of corruption  Index} 
#' \item{Government_Effectiveness}{ Government Effectiveness  Index} 
#' \item{Political_Stability}{ Political Stability  Index} 
#' \item{Regulatory_Quality}{ Regulatory Quality  Index} 
#' \item{Rule_of_Law_2}{ Rule of Law 2  Index} 
#' \item{Voice_and_Accountability}{ Voice and Accountability  Index} 
#' \item{Current_Account_Balance}{ Current Account Balance  Index} 
#' \item{Agricultural_Workers}{ Agricultural Workers  Index} 
#' \item{Self_Employed}{ Self Employed  Index} 
#' \item{Service_Workers}{ Service Workers  Index} 
#' \item{GDP_PP}{ GDP PP  Index} 
#' \item{gini_disp}{ gini disp  Index} 
#' \item{Females_in_Laborforce}{ Females in Laborforce  Index} 
#' \item{Land_Area}{ Land Area  Index} 
#' \item{Female_Life_Expectancy}{ Female Life Expectancy  Index} 
#' \item{Male_Life_Expectancy}{ Male Life Expectancy  Index} 
#' \item{Total_Life_Expectancy}{ Total Life Expectancy  Index} 
#' \item{Female_Literacy}{ Female Literacy  Index} 
#' \item{Male_Literacy}{ Male Literacy  Index} 
#' \item{Total_Literacy}{ Total Literacy  Index} 
#' \item{Defense_Spending}{ Defense Spending  Index} 
#' \item{Population_0_to_14}{ Population 0 to 14  Index} 
#' \item{Population_15_to_64}{ Population 15 to 64  Index} 
#' \item{Population_65_plus}{ Population 65 plus  Index} 
#' \item{Female_Population}{ Female Population  Index} 
#' \item{Total_Population}{ Total Population  Index} 
#' \item{Ratio_Females_to_Males_in_Laborforce}{ Ratio Females to Males in Laborforce  Index} 
#' \item{Rural_Population}{ Rural Population  Index} 
#' \item{Ratio_School_to_Tertiary_Enrolment}{ Ratio School to Tertiary Enrolment  Index} 
#' \item{Tax}{ Tax  Index} 
#' \item{Urban_Population}{ Urban Population  Index} 
#' \item{Happiness}{ Happiness  Index} 
#' \item{Homicide}{ Homicide  Index} 
#' }
"FullCountryDataFrameForPlots"


#' The final data set for archetypal mixture for the 5 archetypes created in WVS wave 6
#'
#' It contains the components of archetypes for the 10 Schwartz values
#'
#' @format A data frame with 5 rows and 11 variables:
#' \describe{
#' \item{B_Archetype}{Alpha label of archetype} 
#' \item{Universalism}{score of Schwartz value} 
#' \item{Benevolence}{score of Schwartz value} 
#' \item{Tradition}{score of Schwartz value} 
#' \item{Conformity}{score of Schwartz value} 
#' \item{Security}{score of Schwartz value} 
#' \item{Power}{score of Schwartz value} 
#' \item{Achievement}{score of Schwartz value} 
#' \item{Hedonism}{score of Schwartz value} 
#' \item{Stimulation}{score of Schwartz value} 
#' \item{Self-direction}{score of Schwartz value} 
#' }
"Final_W6_Archetypes"


#' The complete cases data set of World Value Survey Wave 6
#' after reversing Scwartz values to the scale 
#' "Not at all like me"->1
#' ...
#' "Very much like me"->6
#'
#' @format A data frame with 83526 rows and 14 variables:
#' \describe{
#' \item{id}{identity stamp for each respondent} 
#' \item{desc_gender}{gender of respondent: 0=male, 1=female} 
#' \item{desc_country}{country numeric code} 
#' \item{cyname}{short country name} 
#' \item{Universalism}{name of Schwartz value} 
#' \item{Benevolence}{name of Schwartz value} 
#' \item{Tradition}{name of Schwartz value} 
#' \item{Conformity}{name of Schwartz value} 
#' \item{Security}{name of Schwartz value} 
#' \item{Power}{name of Schwartz value} 
#' \item{Achievement}{name of Schwartz value} 
#' \item{Hedonism}{name of Schwartz value} 
#' \item{Stimulation}{name of Schwartz value} 
#' \item{Self-direction}{name of Schwartz value} 
#' }
"W6"



R/country_analysis.R
#' @title Create country mixtures for archetypes and indices HH, NH, MI
#' @description
#' Read processed data frame, read variables, archetype data frame, variable category and 
#' create country mixtures for archetypes plus indices HH, NH, MI
#' @param wavedatain The rds file with columns the World Values to be used
#' @param varnames R The vector of names of values to be used
#' @param archetypesin The rds file with columns the final archetype found
#' @param varcat The character name of values category (default='Schwartz')
#' @param exclude Logical entry set to FALSE
#' @return NULL
#' @importFrom DescTools Herfindahl
#' @importFrom openxlsx write.xlsx
#' @export
#' @examples 
#' #country_analysis(wavedatain,varnames,archetypesin)
country_analysis <-
function(wavedatain,varnames,archetypesin,varcat='Schwartz',exclude=FALSE){
# country mixture analysis
# now using Wave 6 complete case data and boot averages for archetypes
# this version generates 3 country proportion tables
# 1) based on full data
# 2) based on excluding those > median of distance to nearest archetype
# 3) based on exclduing those closer to "middle" (global average) than nearest archetype
#
# exclude <- FALSE
# library(DescTools)
# library(WriteXLS)
sink(file=paste0(varcat,"_country_analysis_report.txt"),split = T)
cat("Country mixture analysis","\n")
t1=Sys.time()
wavedata <- readRDS(wavedatain)
cat(paste0("summary of input respondent data for ",varcat),"\n")
print(t(apply(wavedata[,varnames],2,summary)),quote=F)
cat("respondents by country","\n")
print(table(wavedata$cyname),quote=F)
cat("Averages of varnames variables are:","\n")
Middle <- apply(wavedata[,varnames],2,mean,na.rm=T)
print(Middle,quote=F)
#######
archetypes <- readRDS(archetypesin) #watch this, can be a list or a simple df
#archetypes <- archetypes[["BestFit_Archetypes"]]
cat("Input archetypes to use","\n")
print(archetypes,quote=F,row.names = F)
####
kappa <- nrow(archetypes)
cat(paste0("Kappa is : ",kappa),"\n")
#
#transpose
t_data <- as.matrix(t(wavedata[,varnames]))
t_arches <- as.matrix(t(archetypes[,varnames]))
###
cat("adding in global average to transposed archetypes","\n")
t_arches <- cbind(t_arches,Middle)
# find nearest archetype
fdistimin=function(XCOL,XC){
  #function that returns the position of minimum distance from the archetypes XC
  ds=colSums((XCOL-XC)^2);return(which.min(ds))}
# distance to nearest archetype
fdistimin2=function(XCOL,XC){
  #function that returns the minimum distance from the archetypes XC
  ds=sqrt(colSums((XCOL-XC)^2));return(min(ds))}
# all k distances returned
fdistall=function(XCOL,XC){
  #function that returns all the k distances from the k archetypes XC
  ds=sqrt(colSums((XCOL-XC)^2));return(ds)}
#
# apply these
nearestA <- apply(t_data,2,fdistimin,XC=t_arches[,1:kappa]) # nearest ignoring global average
nearestA_d <- apply(t_data,2,fdistimin2,XC=t_arches[,1:kappa])
all_d <- apply(t_data,2,fdistall,XC=t_arches) # this includes global average
all_d <- t(all_d) #others are simple vectors, this matrix
colnames(all_d) <- paste0("d_",colnames(all_d))
###
wavedata <- cbind(wavedata,nearestA,nearestA_d,all_d)
#
dist_med <- aggregate(wavedata$nearestA_d,list(A=wavedata$nearestA),median)
colnames(dist_med) [which(colnames(dist_med)=="x")] <- "med_dist"
cat("medians of distances to nearest archetype, by archetype","\n")
print(dist_med,quote = F)
wavedata <- merge(wavedata,dist_med,by.x="nearestA",by.y="A",all.x = T,sort=F)
wavedata[,"Assign_med_d"] <- ifelse(wavedata[,"nearestA_d"] <= wavedata[,"med_dist"],TRUE,FALSE)
wavedata[,"Archetype_med_d"] <- ifelse(wavedata[,"Assign_med_d"],wavedata[,"nearestA"],0)
wavedata[,"Assign_middle"] <- ifelse(wavedata[,"nearestA_d"] <= wavedata[,"d_Middle"],TRUE,FALSE)
wavedata[,"Archetype_middle"] <- ifelse(wavedata[,"Assign_middle"],wavedata[,"nearestA"],0)
#Save processed file
saveRDS(wavedata,paste0("data_nearest.rds"))
#
tab_out <- vector(mode="list",length=3)
names(tab_out) <- c("Full","Median","Middle")
# tablulate full data
tab1 <- table(wavedata$cyname,wavedata$nearestA)
tab2 <- round(prop.table(tab1,margin=1),5)
out <- as.data.frame.matrix(tab2)
colnames(out)  <- row.names(archetypes)
tab_out[[1]] <- out
# tabulate median exclusion
tab1 <- table(wavedata$cyname,wavedata$Archetype_med_d)
tab2 <- round(prop.table(tab1,margin=1),5)
out <- as.data.frame.matrix(tab2)
colnames(out)  <- c("Median_Exclude",row.names(archetypes))
tab_out[[2]] <- out
# tabulate global average exclusion
tab1 <- table(wavedata$cyname,wavedata$Archetype_middle)
tab2 <- round(prop.table(tab1,margin=1),5)
out <- as.data.frame.matrix(tab2)
colnames(out)  <- c("Global_Exclude",row.names(archetypes))
tab_out[[3]] <- out
for (kk in 1:3) {
# herfindahl index
# max is 1 but here we have floor
t_col <- ncol(tab_out[[kk]])
floor <- Herfindahl(rep(1/t_col,t_col))
tab_out[[kk]] [,"HH"] <- round(apply(tab_out[[kk]],1,Herfindahl),2)
# normalize
tab_out[[kk]] [,"NHH"] <- round(((tab_out[[kk]] [,"HH"] - floor)/(1 - floor)),2)
# Sunil's index
tab_out[[kk]] [,"MI"] <- 1 - tab_out[[kk]] [,"NHH"]
print(paste0("Table: ",names(tab_out[kk])),quote=F)
print(round(tab_out[[kk]],2),quote=F)
print(summary(tab_out[[kk]]),quote=F)
temp <- as.data.frame(matrix(nrow=nrow(tab_out[[kk]]),ncol=2))
temp[,1] <- paste0(row.names(tab_out[[kk]]))
temp[,2] <- tab_out[[kk]] [,"MI"]
print(paste0("Table: ",names(tab_out[kk]),"--countries ordered by decreasing MI"),quote=F)
print(temp[order(temp[,2],decreasing = T),],quote=F)
}
# WriteXLS(tab_out,paste0("Kappa_",kappa,"_country_tables.xlsx"),row.names = TRUE)#deprecated due to Perl installation needed
write.xlsx(tab_out,paste0("Kappa_",kappa,"_country_tables.xlsx"),creator='David F. Midgley',row.names=TRUE,col.names=TRUE)
#
sink(file=NULL)
#eof
}


R/impute_variables.R
#' @title Impute requested R variables using package 'mice' in parallel mode
#' @description
#' Read excel file of description for imputing variables, read missing values data frame, impute and save results
#' @param inxlfile Excel file of variables to be imputed
#' @param df R rds file of data with missing values to be imputed
#' @param ncores the number of logical cores that should be used (apply for multicore computers)
#' @return NULL
#' @importFrom readxl read_excel
#' @import parallel
#' @import foreach
#' @import iterators
#' @import doParallel
#' @import lattice
#' @import mice
#' @export
#' @examples
#' #impute_variables_R(inxlfile,df)
impute_variables <-
function(inxlfile,df,ncores=NULL){
  #Function to impute variables given by excel file 'inxlfile'
  #
  ######Check arguments:
  #
  if(all(is.null(c(inxlfile,df)))){
    stop('You must use two files for input.')
  }
  #
  #Check for number of cores
  #
  if(is.null(ncores)){ncores=detectCores()}
  #
  if(ncores>detectCores()){ncores=detectCores()}
  #
  ##########################################################################################################
  ######Read chosen variables and methods for imputation
  ##########################################################################################################
  #
  #Data frame of variables:
  dvars=as.data.frame(read_excel(inxlfile));
  vars=dvars$name
  varsnames=dvars[dvars$name%in%vars,"rename"]
  #Keep only required variables:
  dw=df[,vars]
  colnames(dw)=varsnames
  # Convert variable to factors...
  dvars$factor=sapply(dvars$type,function(x){ifelse(x!="numeric",{return(TRUE)},{return(FALSE)})});
  #
  dvars[dvars$factor,'rename']
  dw[,dvars[dvars$factor,'rename']] <- lapply(dw[,dvars[dvars$factor,'rename']] , factor)
  #Check the percentage of missing values
  yy=apply(dw,2,function(x){100*sum(is.na(x))/length(x)})
  #Use of 'rule of thumb': 1 imputation per average each % of missing values
  nimps=floor(mean(yy,na.rm = T))
  cat(paste0('The rule of thumb criterion: "1 imputation per each 1% of average missing values" suggests ',nimps,' imputations'),'\n')
  # Create vector of proper imputation methods for mice:
  vmethod=dvars$method;vmethod
  #Function for a single imputation:
  fmice=function(i,df,method){mice(df,m=1,method = method)}
  #Do Parallel:
  #Check availability of cores to satisfy the 'rule of thumb' criterion:
  #
  if(ncores==1){
    warning('It seems that there exist only 1 core, so parallel imputation cannot be applied','\n')
    nused=1
  }else if(ncores>nimps){
    nused=nimps
    cat(paste0('There will be used ',nused,' cores for parallel mice imputation'),'\n')
    cat(paste0('The rule of thumb criterion for choosing number of imputations is satisfied'),'\n')
  }else{
    nused=ncores
    cat(paste0('There will be used ',nused,' cores for parallel mice imputation'),'\n')
    cat(paste0('The rule of thumb criterion for choosing number of imputations is not satisfied'),'\n')
  }
  #
  i=NULL
  cl <- makeCluster(nused);registerDoParallel(cl);
  t1=Sys.time();
  m22=foreach(i=1:nused,.combine=ibind,.packages = c('mice')) %dopar% {fmice(i,df=dw,method=vmethod)};
  t2=Sys.time();print(as.POSIXlt(t2, "GMT")-as.POSIXlt(t1, "GMT"),quote=F);#
  stopCluster(cl);
  #
  #Save output:
  #
  fname=paste0('dimp',nused,'.rds')
  saveRDS(m22,fname)
  #
  cat(paste0(dim(dw)[2],' variables were imputed using "mice" with ',nused,' imputations'),'\n')
  cat(paste0('Data frame of imputed variables has been stored in file "',fname,'" as a "mids" object'),'\n')
  #
  }


R/reverse_variables.R
#' @title Reverse requested R variables
#' @description
#' Read excel file of description for reversing variables
#' @param inxlfile Excel file of variables
#' @param dr R rds file of data
#' @param validansR Valid answers file
#' @return NULL
#' @examples
#' #reverse_variables(inxlfile,dr,validansR=validr)
#' @importFrom readxl read_excel
#' @importFrom plotrix rescale
#' @export
reverse_variables <-
function(inxlfile,dr,validansR){
  #Function to reverse and rescale chosen variables from excel file 'inxlfile'
  #
  ######Check arguments:
  #
  if(all(is.null(c(inxlfile,dr,validansR)))){
    stop('You must use three files for input.')
  }
  #
  #
  ##########################################################################################################
  ######Read chosen variables and instructions for reversing and rescaling
  ##########################################################################################################
  #
  dvars=as.data.frame(read_excel('in3.xlsx'),stringsAsFactors=FALSE)
  dvars$factor=as.logical(dvars$factor)
  dvars$reverse=as.logical(dvars$reverse)
  vall=dvars$name
  #All processes variables
  vallr=names(validansR)
  #Factors:
  vars=dvars[dvars$factor,"name"]
  #Find factor variables to be reversed
  vrev=dvars[dvars$reverse,"name"]
  #Find variables to be rescaled
  vch=dvars[!is.na(dvars$rescale),"name"]
  #To be reversed and to be rescaled:
  vars2=c(vrev,vch)
  #Find range for
  vranges=matrix(as.integer(unlist(strsplit(dvars[dvars$name%in%vch,"rescale" ],";"))),ncol=2,byrow = T)
  rownames(vranges)=vch
  colnames(vranges)=c('MIN','MAX')
  #
  #Create data frame of valid answers only:
  #
  #
  dvlist=lapply(vallr, function(vname,validans,df){
    out=df[,vname]
    ii=df[,vname]%in%validans[[vname]][,vname]
    out[ii]=df[ii,vname]
    out[!ii]=NA
    return(out)
  },validans=validansR,df=dr)
  names(dvlist)=vallr
  dv=do.call(cbind,dvlist)
  #
  #
  #Function for reversing values:
  #
  frevcol=function(cn){
    xmin=min(cn,na.rm=TRUE)
    xmax=max(cn,na.rm=TRUE)
    fx=function(x,xmin,xmax){xmax+xmin-x}
    yy=fx(cn,xmin,xmax)
    return(yy)
  }
  #
  #Reverse them:
  #
  if(length(vrev)==1){
    dv[,vrev]=frevcol(dv[,vrev])
  }else {
    dv[,vrev]=apply(dv[,vrev],2,frevcol)
  }
  #
  #
  #Use rescale form 'plotrix' package for those that must be rescaled:
  #
  if(length(vch)==1){
    dv[,vch]=rescale(dv[,vch],vranges)
  }else {
    yy=sapply(vch, function(vn,df,ranges){
      xrange=as.integer(ranges[vn,])
      out=rescale(df[,vn],xrange)
      return(out)
    },df=dv[,vch],ranges=vranges)
    dv[,vch]=yy
  }
  #
  #Store both reversed, rescaled and untouched variables:
  #
  dwr=dr[,vall]
  dwr[,vallr]=dv[,vallr]
  dwr[,vrev]=dv[,vrev]
  dwr[,vch]=dv[,vch]
  #
  #Save working data frame dwr:
  #
  saveRDS(dwr,'dwr.rds')
  #OK
  #
  cat(paste0(length(vall),' variables were chosen'),'\n')
  cat(paste0(length(vrev),' variables were reversed'),'\n')
  cat(paste0(length(vch),' variables were rescaled'),'\n')
  cat(paste0('Data frame of valid answers after reverses and rescales has been stored in file "dwr.rds"'),'\n')
  #
}


R/mds_country_mixtures.R
#' @title Create MDS configuration for countries
#' @description
#' Read country data and parameters in order ot create MDS configuration
#' @param n_arches The integer number of archetypes used
#' @param ndim The integer dimension of MDS configuration
#' @param countrydata The excel file with country data to be used as input
#' @param exclude Logical entry set to FALSE
#' @param dmetric The distance metric to be used
#' @param usebest Logical entry set to FALSE if Torgerson will be used
#' @param perm Logical entry set to TRUE if permutation will be performed
#' @param jack Logical entry set to TRUE if jackknife will be performed
#' @param use_indices Logical entry set to TRUE if country indices will be used
#' @return NULL
#' @importFrom readxl read_excel
#' @importFrom tibble rownames_to_column
#' @importFrom plotrix color.id
#' @import smacof
#' @importFrom robCompositions aDist
#' @importFrom openxlsx write.xlsx
#' @examples
#' #mds_country_mixtures(n_arches,ndim,countrydata)
#' @export
mds_country_mixtures=function(n_arches,ndim=3,countrydata,exclude=FALSE,dmetric="standard",usebest=FALSE,perm=TRUE,jack=TRUE,use_indices=TRUE){
# mapping country mixtures
# this version, switch for altered index numbers if "excluded"
# reading tables from multisheet Excel
run_title <- paste0("MDS_exclude_",exclude,"_dmetric_",dmetric,"_D",ndim)
sink(file=paste0(run_title,".txt"),split=T)
cat("settings are","\n")
cat("archetypes = ",n_arches,"dimensions = ",ndim,"\n")
cat("exclude = ",exclude,"dmetric = ",dmetric,"\n")
cat("use best = ",usebest,"\n")
if(dmetric=="Aitchison"){ cat("Aitchison distance metric used","\n") }else{ cat("Euclidean distance metric used", "\n")}
#
if(dmetric=="standard"){ dmethod <- "euclidean"}
#
if(exclude){ sheetname <- "Median" }else{sheetname <- "Full"}
cat("reading sheet:",sheetname,"\n")
dinput <- as.data.frame(read_excel(paste0("Kappa_",n_arches,"_country_tables.xlsx"),sheet = sheetname))
dinput[,c("HH","NHH","MI")] <- NULL
colnames(dinput) [1] <- "Country"
print(head(dinput),quote=F)
if (exclude){dinput$Median_Exclude <- NULL}# drop them if present
a_names <- colnames(dinput[2:(n_arches+1)]) # read archetypes given names
# get country data
country_data <- as.data.frame(read_excel(countrydata))
dinput <- merge(dinput,country_data,by="Country",all.x = FALSE,sort=FALSE) # keep countries in mixture table/input order
#Remove WVS_Code
dinput=dinput[,setdiff(colnames(dinput),'WVS_Code')]
####
if(dmetric=="standard"){dissim <- dist(dinput[,a_names],diag=TRUE,method = dmethod)}
if(dmetric=="Aitchison"){dissim <- aDist(dinput[,a_names])}
attr(dissim,"Labels") <- dinput[,(n_arches+2)] # watch this, it is the ISO code column
res <- smacofSym(dissim,ndim = ndim,type="ratio")
cat(paste0("overall stress: ",res$stress),"\n")
cat("stress by country: ","\n")
print(res$spp,quote=F)
cat("random starting configurations","\n")
stressvec <- NULL
set.seed(1234567)
fitran <- vector(mode ="list",length=20)
for (k in 1:30) {
  fitran[[k]] <- smacofSym(dissim,ndim=ndim,init="random",type = "ratio")
  stressvec[k] <- fitran[[k]]$stress
}
cat("stress for random starts","\n")
print(stressvec,quote=F)
best <- which.min(stressvec)
cat(paste0("minimum of ",stressvec[best]," found at ",best),"\n")
if(stressvec[best] < res$stress) {
      temp <- round((1 - (stressvec[best]/res$stress))*100,2)
      cat(paste0("random solution is ",temp," % better than classical solution"),"\n")
      if(usebest) {
        cat("random solution is used hereafter","\n")
        res <- fitran[[best]]
        } else cat("however, set to use classical solution", "\n")
} else cat("classical solution is best and used hereafter","\n")
#####
pdf(file=paste0("plots_",run_title,".pdf"),paper="a4r",height=9,width=11)
#
if(perm) {
cat("permutation test of solution:","\n")
res.perm <- permtest(res,nrep = 30,verbose = F)
print(res.perm,quote=F)
hist(res.perm$stressvec,xlab="Stress Values", main="Permutations")
}
#######
if(jack) {
jackfit <- jackknife(res)
cat("jacknife test of stability:","\n")
print(jackfit,quote=F)
}
#####
# output configurations at this point
cat("writing final configurations to .rds & Excel","\n")
configs <- as.data.frame(res$conf)
dnames <- colnames(configs)
configs <- rownames_to_column(configs,var = "ISOAlpha2")
#Ssve configurations:
saveRDS(configs,paste0("configs_",run_title,".rds"))
# WriteXLS(configs,paste0("configurations_",run_title,".xlsx"),row.names = FALSE)
write.xlsx(configs,paste0("configurations_",run_title,".xlsx"),row.names=FALSE,col.names=TRUE,sheetName='configs')
########
cat("biplots of archetypes start:","\n")
x <- biplotmds(res,dinput[,a_names])
print(summary(x))
vs <- 0.7
lim_x <- c(-1.5,1.5)
lim_y <- c(-1.5,1.5)
plot(res,plot.type = "stressplot")
plot(x,plot.dim = c(1,2),main="Country Mixtures + Archetypes",vecscale=vs)
plot(x,plot.dim = c(1,3),main="Country Mixtures + Archetypes",vecscale=vs)
plot(x,plot.dim = c(2,3),main="Country Mixtures + Archetypes",vecscale=vs)
#
if(use_indices) {
  # merge configurations and dinput
  dnext <- merge(dinput,configs,by="ISOAlpha2",sort=FALSE)
  # build table of biplot regressions
  # "dinput" has archetype mixtures, plus indices
  # for plots of indices we need to do one by one, also this allows to use maximum numbers of observations for each index
  table_biplot <- as.data.frame(matrix(nrow=length(c(a_names,9:(ncol(dnext)-3))),ncol=17))
  colnames(table_biplot) <- c("Variable",paste0("coef_D",1:3),paste0("t_D",1:3),paste0("t_prob_D",1:3),
                              "R_Square","F","df1","df2","F_prob","test","include")
  jj <- 1
  X <- as.matrix(dnext[,dnames]) #used in all cases, order enforced
  for (ii in c(3:7,9:50)) {
    a_biplot <- summary(lm(scale(dnext[,ii]) ~ -1 + X))
    table_biplot[jj,"Variable"] <- colnames(dnext) [ii]
    table_biplot[jj,"coef_D1"] <- a_biplot$coefficients[1,"Estimate"]
    table_biplot[jj,"coef_D2"] <- a_biplot$coefficients[2,"Estimate"]
    table_biplot[jj,"coef_D3"] <- a_biplot$coefficients[3,"Estimate"]
    table_biplot[jj,"t_D1"] <- a_biplot$coefficients[1,"t value"]
    table_biplot[jj,"t_prob_D1"] <- a_biplot$coefficients[1,"Pr(>|t|)"]
    table_biplot[jj,"t_D2"] <- a_biplot$coefficients[2,"t value"]
    table_biplot[jj,"t_prob_D2"] <- a_biplot$coefficients[2,"Pr(>|t|)"]
    table_biplot[jj,"t_D3"] <- a_biplot$coefficients[3,"t value"]
    table_biplot[jj,"t_prob_D3"] <- a_biplot$coefficients[3,"Pr(>|t|)"]
    table_biplot[jj,"R_Square"] <- a_biplot$r.squared
    table_biplot[jj,"F"] <- a_biplot$fstatistic["value"]
    table_biplot[jj,"df1"] <- a_biplot$fstatistic["numdf"]
    table_biplot[jj,"df2"] <- a_biplot$fstatistic["dendf"]
    table_biplot[jj,"F_prob"] <- pf(a_biplot$fstatistic["value"],
                                    a_biplot$fstatistic["numdf"],a_biplot$fstatistic["dendf"],lower.tail = F)
    if(table_biplot[jj,"F_prob"] < 0.001) {
      table_biplot[jj,"test"] <- "p<0.001"
      table_biplot[jj,"include"] <- TRUE
      } else {
      if(table_biplot[jj,"F_prob"] < 0.005) {
        table_biplot[jj,"test"] <- "p<0.005"
        table_biplot[jj,"include"] <- TRUE
      } else {
          table_biplot[jj,"test"] <- "ns"
          table_biplot[jj,"include"] <- FALSE
          }}
    jj <- jj + 1
    }
  table_biplot[,c(2:12,15)] <- round(table_biplot[,c(2:12,15)],3)
  # WriteXLS(table_biplot,paste0("biplot_regressions_",run_title,".xlsx"))
  write.xlsx(table_biplot,paste0("biplot_regressions_",run_title,".xlsx"),row.names=FALSE,col.names=TRUE,sheetName='table_biplot')
}
#
dev.off()
sink(file=NULL)
#eof
}
